src/controllers/aer_controller.hpp

```diff
@@ -1497,92 +1438,206 @@ void Controller::run_single_shot(const Circuit &circ, State_t &state,
+
+  bool can_sample = circ.can_sample;
+
   // Optimize circuit
   Noise::NoiseModel dummy_noise;
 
   auto fusion_pass = transpile_fusion(method, circ.opset(), config);
   fusion_pass.optimize_circuit(circ, dummy_noise, state.opset(), result);
 
-  // Check if measure sampling supported
-  const bool can_sample = check_measure_sampling_opt(circ, method);
-  
   // Cache blocking pass
-  uint_t block_bits = 0;
-  if (cache_blocking) {
-    auto cache_block_pass = transpile_cache_blocking(method, circ, dummy_noise, config);
-    cache_block_pass.set_sample_measure(can_sample);
-    cache_block_pass.optimize_circuit(circ, dummy_noise, state.opset(), result);
-    if (cache_block_pass.enabled()) {
-      block_bits = cache_block_pass.block_bits();
-    }
+  uint_t block_bits = circ.num_qubits;
+  auto cache_block_pass = transpile_cache_blocking(method, circ, dummy_noise, config);
+  cache_block_pass.set_sample_measure(can_sample);
+  cache_block_pass.optimize_circuit(circ, dummy_noise, state.opset(), result);
+  if (cache_block_pass.enabled()) {
+    block_bits = cache_block_pass.block_bits();
   }
+  // Check if measure sampling supported
+  can_sample &= check_measure_sampling_opt(circ, method);
 
-  // allocate qubit register
-  state.allocate(max_qubits_, block_bits);
-
-    // Check if measure sampler and optimization are valid
+  // Check if measure sampler and optimization are valid
   if (can_sample) {
     // Implement measure sampler
     auto& ops = circ.ops;
     auto first_meas = circ.first_measure_pos; // Position of first measurement op
     bool final_ops = (first_meas == ops.size());
 
+    state.set_max_matrix_qubits(get_max_matrix_qubits(circ) );
+
+    // allocate qubit register
+    state.allocate(circ.num_qubits, block_bits);
+
     // Run circuit instructions before first measure
     state.initialize_qreg(circ.num_qubits);
     state.initialize_creg(circ.num_memory, circ.num_registers);
+
+    RngEngine rng;
+    rng.set_seed(circ.seed);
     state.apply_ops(ops.cbegin(), ops.cbegin() + first_meas, result, rng, final_ops);
 
     // Get measurement operations and set of measured qubits
-    measure_sampler(circ.ops.begin() + first_meas, circ.ops.end(), shots, state, result, rng);
+    measure_sampler(circ.ops.begin() + first_meas, circ.ops.end(), circ.shots, state, result, rng);
 
     // Add measure sampling metadata
     result.metadata.add(true, "measure_sampling");
-  } else {
+  }
+  else{
     // Perform standard execution if we cannot apply the
     // measurement sampling optimization
-    while (shots-- > 0) {
-      run_single_shot(circ, state, result, rng);
+
+    if(enable_batch_multi_shots_ && !cache_block_pass.enabled()){
+      //apply batched multi-shots optimization on GPU
+      Base::MultiStates<State_t> states;
+
+      states.set_parallelization(max_batched_states_);
+
+      states.set_max_matrix_qubits(get_max_matrix_qubits(circ) );
+
+      states.allocate(circ.num_qubits, circ.num_qubits,circ.shots);
```

Comment:
This second argument (`circ.num_qubits`) is different from others (`block_bits`). Is this intentional?